### DISCUSSION TITLE
[rocblas] Tensile doesn't adhere to requested parallelism

### DISCUSSION THREAD

**Discussion by wjhorne** [rocm-arch/rocm-arch#783] [Ideas]
Hi,

I ran into an issue where building rocmblas was using far more threads than the NINJAFLAGS variable should allow. Each thread uses ~5GB ram which creates kind of an insane amount of ram usage (96 GB or so in total for my machine) making the build completely impossible on my machine. 

I found a relatively simple fix for it. If you add -DTensile_CPU_THREADS="$NUM_THREADS" to the cmake input deck for rocmblas you can limit the threads of the build to sane levels. I would have opened a PR with it, but I wasn't sure of the best path to set NUM_THREADS. Maybe just parsing NINJAFLAGS?

**Comment by JP-Ellis** [comment]
This issue might be worth reporting upstream too to see what they think.

As to the implementation, I think that parsing flags would be a difficult option to implement (there's a reason why there are entire libraries dedicated to parsing flags). Furthermore, which flags should be parsed?  `NINJAFLAGS` is unlikely to be set at all as `ninja` is one of the less common build system, so it might be better to try and parse `MAKEFLAGS` for example.

As for the best way to get the number of threads, this would be `nproc`, so something like this would work:

```sh
-DTensile_CPU_THREADS=$(nproc)
```

Lastly, you say the issue is primarily to do with RAM.  So should we instead have a heuristic based on RAM instead of the number of CPUs?  So it might be easy to use the RAM and see how many times we can fit 5GiB:

```sh
NUM_THREADS=$(grep MemTotal /proc/meminfo | awk '{print int($2 / 1024 / 1024 / 5)}')
```

**Comment by lahwaacz** [comment]
Note that `NINJAFLAGS` does not work the same way as `MAKEFLAGS`: the latter is a standard env variable parsed directly by `make`, but `ninja` does not parse any env variable and all parameters have to be passed on the command line. Some PKGBUILDs like rocm-llvm use `NINJAFLAGS` via `ninja $NINJAFLAGS` to make the flags configurable by users. The `rocblas` PKGBUILD calls `make`, not `ninja`.

**Comment by acxz** [comment]
@wjhorne If you end up making an upstream issue please link it to this discussion. If there ends up being a patch for it, feel free on making a PR to add that in our PKGBUILD.

**Reply by wjhorne** [reply]
I'll make sure to do so if I open up something upstream or otherwise. Apologies for dropping off the map. I got pulled into some other work.
